src/main/java/duke/Duke.java

```diff
@@ -1,91 +1,140 @@
         storage.updateData(this.taskList);
-        ui.displayMessage("Bye. Hope to see you again soon!\n");
+        ui.displayGoodbyeMessage();
         System.exit(0);
     }
 
+    private String taskAddedMessage(Task task) {
+        return "Got it. I've added this task:\n" +
+                task.toString() +
+                "\nNow you have " +
+                taskList.size() +
+                " tasks in the list\n";
+    }
+
     /**
-     * Starts an instance of the chatbot by displaying welcome message and beginning to listen for user input.
+     * Returns a <code>Message</code> that represents the response of the chatbot when prompted by an input by user.
+     * @param msg A <code>Message</code> object representing the user's request
+     * @return A <code>Message</code> object containing the response of the bot to the user's request
      */
-    void run() {
-        ui.welcomeMessage();
-        while (true) {
-            String[] tokens = parser.parseUserInput();
-            if (tokens.length == 1 && tokens[0].equals("bye")) {
-                exit();
-                break;
-            } else if (tokens.length == 1 && tokens[0].equals("list")) {
-                ui.displayItemList(taskList);
-            } else if (tokens[0].equals("mark")) {
-                taskList.markListItem(tokens, ui);
-            } else if (tokens[0].equals("unmark")) {
-                taskList.unmarkListItem(tokens, ui);
-            } else if (tokens[0].equals("todo")) {
-                try {
-                    taskList.addToDo(tokens, ui);
-                } catch (DukeException e) {
-                    ui.displayMessage(e.getMessage());
-                }
-            } else if (tokens[0].equals("deadline")) {
-                try {
-                    taskList.addDeadline(tokens, ui);
-                } catch (DukeException e) {
-                    ui.displayMessage(e.getMessage());
-                }
-            } else if (tokens[0].equals("event")) {
-                try {
-                    taskList.addEvent(tokens, ui);
-                } catch (DukeException e) {
-                    ui.displayMessage(e.getMessage());
-                }
-            } else if (tokens[0].equals("delete")) {
-                try {
-                    taskList.deleteItem(tokens, ui);
-                } catch (DukeException e) {
-                    ui.displayMessage(e.getMessage());
-                }
-            } else if (tokens[0].equals("find")) {
-                try {
-                    taskList.findItemInList(tokens, ui);
-                } catch (DukeException e) {
-                    ui.displayMessage(e.getMessage());
-                }
-            } else {
-                ui.displayMessage("unknown command\n");
+    public Message respondToMessage(Message msg) {
+        String input = msg.getMessage();
+        String[] tokens = parser.parseUserInput(input);
+
+        if (tokens.length == 1 && tokens[0].equals("bye")) {
```

Comment:
I think there is potential to reduce the clutter in the Duke class, perhaps you could refactor them with a class for each command and handle inputs in the parser (which would call the commands separately)?